File: code/890-find-and-replace-pattern.go
package main

import "fmt"

func findAndReplacePattern(words []string, pattern string) []string {
	res := []string{}
	for _, word := range words {
		if len(word) != len(pattern) {
			continue
		}
		m1 := make(map[byte]byte)
		m2 := make(map[byte]byte)
		valid := true
		for i := 0; i < len(pattern); i++ {
			a, b := word[i], pattern[i]
			va, ok_a := m1[a]
			vb, ok_b := m2[b]
			if ok_a != ok_b {
				valid = false
				break
			} else if ok_a && ok_b && !(va == b && vb == a) {
				valid = false
				break
			} else {
				m1[a] = b
				m2[b] = a
			}
		}
		if valid {
			res = append(res, word)
		}
	}
	return res
}

func main() {
	words := []string{"abc", "deq", "mee", "aqq", "dkd", "ccc"}
	pattern := "abb"
	fmt.Println(findAndReplacePattern(words, pattern))
}

File: code/890-find-and-replace-pattern.py
from typing import List


class Solution:
    def findAndReplacePattern(
        self, words: List[str], pattern: str
    ) -> List[str]:
        res = []
        for word in words:
            if len(word) != len(pattern):
                continue
            m1 = {}
            m2 = {}
            valid = True
            for i in range(len(pattern)):
                a, b = word[i], pattern[i]
                va = m1.get(a)
                vb = m2.get(b)
                if (va is None and vb is not None) or (
                    va is not None and vb is None
                ):
                    valid = False
                    break
                elif (
                    va is not None
                    and vb is not None
                    and not (va == b and vb == a)
                ):
                    valid = False
                    break
                else:
                    m1[a], m2[b] = b, a
            if valid:
                res.append(word)
        return res


if __name__ == "__main__":
    words = ["abc", "deq", "mee", "aqq", "dkd", "ccc"]
    pattern = "abb"
    print(Solution().findAndReplacePattern(words, pattern))
